File: spring-web-java/src/test/java/spring/web/java/domain/item/serivce/ItemServiceTest.java
package spring.web.java.domain.item.serivce;

import static org.assertj.core.api.Assertions.*;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.context.annotation.Import;

import spring.web.java.domain.item.Item;
import spring.web.java.domain.item.dto.ItemResponse;
import spring.web.java.domain.item.dto.ItemSaveRequest;
import spring.web.java.domain.item.repository.ItemRepository;
import spring.web.java.global.TestConfig;

@DataJpaTest
@Import(TestConfig.class)
class ItemServiceTest {

	private final ItemRepository itemRepository;
	private final ItemService itemService;

	@Autowired
	public ItemServiceTest(ItemRepository itemRepository) {
		this.itemRepository = itemRepository;
		this.itemService = new ItemService(itemRepository);
	}

	@Test
	void findItem() {
		// Given
		Item item = Item.create("이름", "설명", 15000, 10, Item.Category.ROLE_BOOK);
		Long itemId = itemRepository.save(item).getId();

		// When
		ItemResponse findItem = itemService.findItem(itemId);

		// Then
		assertThat(findItem.name()).isEqualTo(item.getName());
		assertThat(findItem.description()).isEqualTo(item.getDescription());
		assertThat(findItem.price()).isEqualTo(item.getPrice());
		assertThat(findItem.quantity()).isEqualTo(item.getQuantity());
	}

	@Test
	void putItem() {
		// Given
		Item item = Item.create("이름", "설명", 15000, 10, Item.Category.ROLE_BOOK);
		Long itemId = itemRepository.save(item).getId();

		// When
		ItemSaveRequest request = new ItemSaveRequest("이름2", "설명2", 20000, 20, Item.Category.ROLE_TICKET);
		itemService.putItem(itemId, request);

		// Then
		Item findItem = itemRepository.findById(itemId).orElseThrow(IllegalArgumentException::new);

		assertThat(findItem.getName()).isEqualTo(request.name());
		assertThat(findItem.getDescription()).isEqualTo(request.description());
		assertThat(findItem.getPrice()).isEqualTo(request.price());
		assertThat(findItem.getQuantity()).isEqualTo(request.quantity());
		assertThat(findItem.getCategory()).isEqualTo(request.category());
	}
}

File: spring-web-kotlin/src/test/kotlin/spring/web/kotlin/domain/item/service/ItemServiceTest.kt
package spring.web.kotlin.domain.item.service

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.context.annotation.Import
import org.springframework.data.repository.findByIdOrNull
import org.springframework.http.HttpStatus
import spring.web.kotlin.domain.item.Item
import spring.web.kotlin.domain.item.dto.ItemSaveRequest
import spring.web.kotlin.domain.item.repository.ItemRepository
import spring.web.kotlin.global.TestConfig
import spring.web.kotlin.global.common.ResponseMessage
import spring.web.kotlin.global.exception.DomainException

@DataJpaTest
@Import(TestConfig::class)
class ItemServiceTest(
    private val itemRepository: ItemRepository,
    private val itemService: ItemService = ItemService(itemRepository)
) : DescribeSpec({
    describe("findItem 메서드는") {
        context("데이터가 없을 때") {
            it("DomainException 예외 발생 한다") {
                shouldThrow<DomainException> { itemService.findItem(0) }.apply {
                    httpStatus shouldBe HttpStatus.NOT_FOUND
                    responseMessage shouldBe ResponseMessage.DATA_NOT_FOUND
                }
            }
        }

        context("데이터가 있을 때") {
            val item = itemRepository.save(Item.create("이름", "설명", 100, 1, Item.Category.ROLE_BOOK))

            it("조회 되어야 한다") {
                itemService.findItem(item.id!!).apply {
                    name shouldBe item.name
                    description shouldBe item.description
                    price shouldBe item.price
                    quantity shouldBe item.quantity
                }
            }
        }
    }

    describe("putItem 메서드는") {
        context("데이터가 없을 때") {
            it("저장 되어야 한다") {
                val request = ItemSaveRequest("이름", "설명", 100, 1, Item.Category.ROLE_BOOK)
                itemService.putItem(1, request)

                itemRepository.findByIdOrNull(1)?.apply {
                    name shouldBe request.name
                    description shouldBe request.description
                    price shouldBe request.price
                    quantity shouldBe request.quantity
                }
            }
        }

        context("데이터가 있을 때") {
            val savedItem = itemRepository.save(Item.create("이름", "설명", 100, 1, Item.Category.ROLE_BOOK))

            it("수정 되어야 한다") {
                val request = ItemSaveRequest("이름2", "설명2", 200, 2, Item.Category.ROLE_TICKET)
                itemService.putItem(savedItem.id!!, request)

                itemRepository.findByIdOrNull(savedItem.id)?.apply {
                    name shouldBe request.name
                    description shouldBe request.description
                    price shouldBe request.price
                    quantity shouldBe request.quantity
                }
            }
        }
    }
})
